fadecandy_driver/src/fadecandy_driver.cpp

```diff
@@ -0,0 +1,300 @@
+//
+// Copyright (c) 2021 Eurotec
+//
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+#include <iomanip>
+#include <iostream>
+#include <libusb-1.0/libusb.h>
+#include <list>
+#include <math.h>
+#include <vector>
+
+#include "./fadecandy_driver.h"
+
+namespace fadecandy_driver
+{
+std::vector<unsigned char> intToCharArray(unsigned char* buffer, int in, const size_t byts_per_int)
+{
+  std::vector<unsigned char> char_array;
+  for (size_t i = 0; i < byts_per_int; i++)
+  {
+    size_t shift = 8 * (byts_per_int - 1 - i);
+    buffer[i] = (in >> shift) & 0xff;
+    char_array.push_back(buffer[i]);
+  }
+  std::reverse(char_array.begin(), char_array.end());
+  return char_array;
+}
+
+void FadecandyDriver::findUsbDevice()
+{
+  libusb_device** list = NULL;
+  int rc = 0;
+  unsigned count = 0;
+
+  rc = libusb_init(&context_);
+
+  count = libusb_get_device_list(context_, &list);
+  for (size_t idx = 0; idx < count; ++idx)
+  {
+    libusb_device* device = list[idx];
+    libusb_device_descriptor desc = { 0 };
+
+    rc = libusb_get_device_descriptor(device, &desc);
+    assert(rc == 0);
+    if (desc.idVendor == USB_VENDOR_ID && desc.idProduct == USB_PRODUCT_ID)
+    {
+      fadecandy_device_ = device;
+      fadecandy_device_descriptor_ = desc;
+    }
+  }
+}
+
+std::vector<std::vector<unsigned char>>
+FadecandyDriver::makeVideoUsbPackets(std::vector<std::vector<Color>> led_array_colors)
+{
+  int led_index = 0;
+  std::vector<Color> all_led_colors(LEDS_PER_STRIP * NUM_STRIPS, { 0, 0, 0 });
+  for (size_t i = 0; i < led_array_colors.size(); i++)
+  {
+    for (size_t j = 0; j < led_array_colors[i].size(); j++)
+    {
+      led_index = (i * LEDS_PER_STRIP) + j;
+      all_led_colors[led_index] = led_array_colors[i][j];
+    }
+  }
+  assert(all_led_colors.size() == LEDS_PER_STRIP * NUM_STRIPS);
+  std::vector<std::vector<unsigned char>> packets;
+  std::vector<unsigned char> packet;
+  std::vector<Color> packet_leds;
+  std::vector<Color> remaining_leds = all_led_colors;
+  std::vector<int> color_bytes;
+  int control;
+
+  while (remaining_leds.size() > 0)
+  {
+    color_bytes.clear();
+    packet.clear();
+
+    if (remaining_leds.size() < LEDS_PER_PACKET)
+    {
+      packet_leds.assign(remaining_leds.begin(), remaining_leds.end());
+      remaining_leds.erase(remaining_leds.begin(), remaining_leds.end());
+    }
+    else
+    {
+      packet_leds.assign(remaining_leds.begin(), remaining_leds.begin() + LEDS_PER_PACKET);
+      remaining_leds.assign(remaining_leds.begin() + LEDS_PER_PACKET, remaining_leds.end());
```

Comment:
Why not use `all_led_colors` and erase the first X each time you loop? Or I suppose you could also rewrite this to a for loop with a step size equal to `LEDS_PER_PACKET` which will eliminate the need for erasing what you parsed each time.

Reply:
I kept as it's and instead of rewriting it I just erased the `LEDS_PER_PACKET` every loop